XenModel/Actions/Updates/UnzipXenServerPatchAction.cs

```diff
@@ -0,0 +1,123 @@
+        {
+            zippedUpdateFilePath = zippedUpdate;
+            updateFileExtension = fileExtension;
+        }
+
+        public string UnzippedUpdatePatchPath { get; private set; }
+
+        private void ExtractFile()
+        {
+            ArchiveIterator iterator = null;
+            try
+            {
+                using (Stream stream = new FileStream(zippedUpdateFilePath, FileMode.Open, FileAccess.Read))
+                {
+                    iterator = ArchiveFactory.Reader(ArchiveFactory.Type.Zip, stream);
+                    DotNetZipZipIterator zipIterator = iterator as DotNetZipZipIterator;
+                    if (zipIterator != null)
+                    {
+                        zipIterator.CurrentFileExtractProgressChanged +=
+                            archiveIterator_CurrentFileExtractProgressChanged;
+                    }
+                    
+                    while (iterator.HasNext())
+                    {
+                        string currentExtension = Path.GetExtension(iterator.CurrentFileName());
+                        if (currentExtension == updateFileExtension.ToLowerInvariant() || currentExtension == ".iso")
+                        {
+                            string path = Path.Combine(Path.GetTempPath(), iterator.CurrentFileName());
+                            log.DebugFormat(
+                                "Found '{0}' in the downloaded archive when looking for a '{1}' file. Extracting...",
+                                iterator.CurrentFileName(), currentExtension);
+                            using (Stream outputStream = new FileStream(path, FileMode.Create))
+                            {
+                                iterator.ExtractCurrentFile(outputStream);
+                                UnzippedUpdatePatchPath = path;
+
+                                log.DebugFormat("Update file extracted to '{0}'", path);
+
+                                break;
+                            }
+                        }
+                    }
+                    if (zipIterator != null)
+                    {
+                        zipIterator.CurrentFileExtractProgressChanged -=
+                            archiveIterator_CurrentFileExtractProgressChanged;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                log.ErrorFormat("Exception occurred when extracting downloaded archive: {0}", ex.Message);
+                throw new Exception(Messages.DOWNLOAD_AND_EXTRACT_ACTION_EXTRACTING_ERROR);
+            }
+            finally
+            {
+                if (iterator != null)
+                    iterator.Dispose();
+            }
+        }
+
+        void archiveIterator_CurrentFileExtractProgressChanged(object sender, ExtractProgressChangedEventArgs e)
+        {
+            int pc = 95 + (int)(5.0 * e.BytesTransferred / e.TotalBytesToTransfer);

```

Comment:
This would report the PercentComplete from 95 to 100%. I think we neded something like this:  `pc = (int)(100.0 * e.BytesTransferred / e.TotalBytesToTransfer)`